src/storage/pooled_storage_manager.h

```diff
@@ -70,7 +71,8 @@ class GPUPooledStorageManager final : public StorageManager {
 
  private:
   void DirectFreeNoLock(Storage::Handle handle) {
-    cudaError_t err = cudaFree(handle.GetDptr());
+    MemoryManager mm;
```

Comment:
I think mm should be defined as a private variable of pooled_storage_manger instead of initializing one every time. Although it might runs fine now because you are directly calling cuda in mm, when you use your own implementation having one new instance every time can be problematic.